Bug: parse the wrong ServerName from addr

https://github.com/quic-go/quic-go/blob/f633dca488b0858088afae0ba4709b3bfdf798f9/transport.go#L176-L184

Test case:
```go
// env:
//  go1.21.0
// github.com/quic-go/quic-go v0.38.0
func TestQUICTLS(t *testing.T) {
	quicConf := &quic.Config{
		HandshakeIdleTimeout:    5 * time.Second,
		DisablePathMTUDiscovery: true,
	}

	tlsConf := &tls.Config{
		NextProtos: []string{"h3"},
	}

	//tlsConf.ServerName = "quic.nginx.org."

	addr := "quic.nginx.org:443"

	_, err := quic.DialAddr(context.TODO(), addr, tlsConf, quicConf)
	if err != nil {
		t.Error(err)
	}

}
```

The result is:
```go
--- FAIL: TestQUICTLS (0.64s)
    example_test.go:30: CRYPTO_ERROR 0x12a (local): tls: failed to verify certificate: x509: certificate is valid for quic.nginx.org, not quic.nginx.org:443
FAIL
exit status 1
```

one way to fix it:
```go
if tlsConf.ServerName == "" {
	if i := strings.LastIndex(hostname, ":"); i > 0 {
		tlsConf.ServerName = hostname[:i]
	} else {
		tlsConf.ServerName = hostname
	}
}
```

--- a/transport.go
+++ b/transport.go
@@ -159,7 +159,7 @@ func (t *Transport) DialEarly(ctx context.Context, addr net.Addr, tlsConf *tls.C
 	return t.dial(ctx, addr, "", tlsConf, conf, true)
 }
 
-func (t *Transport) dial(ctx context.Context, addr net.Addr, hostname string, tlsConf *tls.Config, conf *Config, use0RTT bool) (EarlyConnection, error) {
+func (t *Transport) dial(ctx context.Context, addr net.Addr, host string, tlsConf *tls.Config, conf *Config, use0RTT bool) (EarlyConnection, error) {
 	if err := validateConfig(conf); err != nil {
 		return nil, err
 	}
@@ -173,15 +173,7 @@ func (t *Transport) dial(ctx context.Context, addr net.Addr, hostname string, tl
 	}
 	tlsConf = tlsConf.Clone()
 	tlsConf.MinVersion = tls.VersionTLS13
-	// If no ServerName is set, infer the ServerName from the hostname we're connecting to.
-	if tlsConf.ServerName == "" {
-		if hostname == "" {
-			if udpAddr, ok := addr.(*net.UDPAddr); ok {
-				hostname = udpAddr.IP.String()
-			}
-		}
-		tlsConf.ServerName = hostname
-	}
+	setTLSConfigServerName(tlsConf, addr, host)
 	return dial(ctx, newSendConn(t.conn, addr, packetInfo{}, utils.DefaultLogger), t.connIDGenerator, t.handlerMap, tlsConf, conf, onClose, use0RTT)
 }
 
@@ -477,4 +469,23 @@ func (t *Transport) ReadNonQUICPacket(ctx context.Context, b []byte) (int, net.A
 	case <-t.listening:
 		return 0, nil, errors.New("closed")
 	}
+}
+
+func setTLSConfigServerName(tlsConf *tls.Config, addr net.Addr, host string) {
+	// If no ServerName is set, infer the ServerName from the host we're connecting to.
+	if tlsConf.ServerName != "" {
+		return
+	}
+	if host == "" {
+		if udpAddr, ok := addr.(*net.UDPAddr); ok {
+			tlsConf.ServerName = udpAddr.IP.String()
+			return
+		}
+	}
+	h, _, err := net.SplitHostPort(host)
+	if err != nil { // This happens if the host doesn't contain a port number.
+		tlsConf.ServerName = host
+		return
+	}
+	tlsConf.ServerName = h
 }

--- a/transport_test.go
+++ b/transport_test.go
@@ -396,6 +396,18 @@ var _ = Describe("Transport", func() {
 		close(packetChan)
 		tr.Close()
 	})
+
+	remoteAddr := &net.UDPAddr{IP: net.IPv4(1, 3, 5, 7), Port: 1234}
+	DescribeTable("setting the tls.Config.ServerName",
+		func(expected string, conf *tls.Config, addr net.Addr, host string) {
+			setTLSConfigServerName(conf, addr, host)
+			Expect(conf.ServerName).To(Equal(expected))
+		},
+		Entry("uses the value from the config", "foo.bar", &tls.Config{ServerName: "foo.bar"}, remoteAddr, "baz.foo"),
+		Entry("uses the hostname", "golang.org", &tls.Config{}, remoteAddr, "golang.org"),
+		Entry("removes the port from the hostname", "golang.org", &tls.Config{}, remoteAddr, "golang.org:1234"),
+		Entry("uses the IP", "1.3.5.7", &tls.Config{}, remoteAddr, ""),
+	)
 })
 
 type mockSyscallConn struct {

